content/blog/2020-03-dplyr-1-0-0-and-vctrs/index.Rmarkdown

```diff
@@ -0,0 +1,149 @@
+  frames.
+* `tidyr::pivot_longer()` has to combine multiple columns into one.
+
+Our goal is to unify the code that underlies all these various functions so that there's one consistent, principled approach. We've already made the change in [tidyr](https://www.tidyverse.org/blog/2019/09/tidyr-1-0-0/), and now it's dplyr's turn. 
+
+## Base R behaviour
+
+You might wonder why we can't just copy the behaviour of `c()`. Unfortunately `c()` has some major downsides:
+
+*   It converts factors to their underlying integer levels:
+    
+    ```{r}
+    c(factor("x"), factor("y"))
+    ```
+  
+*   It doesn't yield sensible results when combining dates and times:
+
+    ```{r}
+    today <- as.Date("2020-03-24")
+    now <- as.POSIXct("2020-03-24 10:34")
+    c(today, now)
+    c(now, today)
+    ```
+
+Additionally, `c()` isn't the only way that base R combines together. `rbind()` and `unlist()` can also be used to perform a similar job, but return different results. This is not to claim that the tidyverse has been any better in the past - we have used a variety of ad hoc method, undoubtedly using well more than three different approaches. 
+
+So decided to come up with our own equivalent of `c()` that we call `vectrs::vec_c()`. `vec_c()`'s behaviour is governed by two main principles:
+
+* Symmetry: `vec_c(x, y)` should return a type as similar as possible to 
+  `vec_c(y, x)`.
+
+* Decrease constraints: this is a bit harder to explain, but if you combine
+  two vectors where one type can represent a subset of the values of the other,
+  you should return the less constrained type. This principle implies that
+  combine an integer and double should return a double, and that combining
+  a date and date-time should return a date-time.
+
+* Prefer to be explicit: if there's confusion about what the result should be,
+  require that the user perform an explicit coercion.
+
+This function resolves the greatest downsides of `c()` and provides a principled foundation for us to build tidyverse functions on top of:
+
+```{r}
+library(vctrs)
+vec_c(today, now)
+vec_c(factor("x"), factor("y"))
+```
+
+## Errors
+
+As a data scientist, you don't really need to know anything about the vctrs package, except that it exists and its used internally by dplyr. (As a software engineer, you might want to learn about vctrs [makes it easier to create new types of vector](https://vctrs.r-lib.org/articles/s3-vector.html)). But vctrs is responsible for creating a number of error messages in dplyr, so it's worth understanding there basic form.
+
+In this first example, we attempt to bind two data frames together where the columns have incompatible types: double and character.
+
+```{r, error = TRUE}
+library(dplyr)
+
+df1 <- tibble(a = 1, b = 1)
+df2 <- tibble(a = 2, b = "a")
+bind_rows(df1, df2)
+```
+Note the components of the error message:
+
+* "No common type" means that vctrs can't find a vector type that can
```

Comment:
This doesn't seem to be part of the error anymore. I'm surprised because we haven't solved r-lib/vctrs#922 yet, but I guess dplyr overrides the messages?

Reply:
Where are you looking? I see

```
#> Error: No common type for `..1$b` <double> and `..2$b` <character>.
```

Reply:
https://deploy-preview-397--tidyverse-org.netlify.com/blog/2020/03/dplyr-1-0-0-and-vctrs/

```
bind_rows(df1, df2)
#> Error: Column `b` can't be converted from numeric to character
```

Reply:
<img width="722" alt="Screenshot 2020-03-25 at 1 05 54 PM" src="https://user-images.githubusercontent.com/4196/77570233-6a424100-6e99-11ea-92ae-fc00ee2bd1ca.png">

Reply:
🤔